sklearn/decomposition/online_lda.py

```diff
@@ -0,0 +1,668 @@
+"""
+
+=============================================================
+Online Latent Dirichlet Allocation with variational inference
+=============================================================
+
+This implementation is modified from Matthew D. Hoffman's onlineldavb code
+Link: http://www.cs.princeton.edu/~mdhoffma/code/onlineldavb.tar
+"""
+
+# Author: Chyi-Kwei Yau
+# Author: Matthew D. Hoffman (original onlineldavb implementation)
+
+import numpy as np
+import scipy.sparse as sp
+from scipy.special import gammaln
+
+from ..base import BaseEstimator, TransformerMixin
+from ..utils import (check_random_state, check_array,
+                     gen_batches, gen_even_slices)
+from ..utils.validation import NotFittedError
+
+from ..externals.joblib import Parallel, delayed, cpu_count
+from ..externals.six.moves import xrange
+
+from ._online_lda import (mean_change, _dirichlet_expectation_1d,
+                          _dirichlet_expectation_2d)
+
+
+def check_non_negative(X, whom):
+    """
+    make sure no Negative data in X
+    (copy from NMF directly)
+    """
+    X = X.data if sp.issparse(X) else X
+    if (X < 0).any():
+        raise ValueError("Negative values in data passed to %s" % whom)
+
+
+def _dirichlet_expectation(X):
+    """
+    For an array theta ~ Dir(X), computes `E[log(theta)]` given X.
+
+    Parameters
+    ----------
+    X : array-like
+        1 or 2 dimention vector
+
+    Returns
+    -------
+    dirichlet_expect: array-like
+        Dirichlet expectation of input array X
+    """
+
+    if len(X.shape) == 1:
+        dirichlet_expect = _dirichlet_expectation_1d(X)
+    else:
+        dirichlet_expect = _dirichlet_expectation_2d(X)
+    return dirichlet_expect
+
+
+def _update_doc_distribution(X, exp_topic_word_distr, doc_topic_prior, max_iters,
+                             mean_change_tol, cal_diff):
+    """
+    E-step: update document topic distribution.
+    (In the literature, this is called `gamma`.)
+
+    Parameters
+    ----------
+    X : sparse matrix, shape=(n_samples, n_features)
+        Document word matrix.
+
+    exp_topic_word_distr : dense matrrix, shape=(n_topics, n_features)
+        Exponential value of expection of log topic word distribution.
+        In the literature, this is `exp(E[log(beta)])`.
+
+    doc_topic_prior : float
+        Prior of document topic distribution `theta`.
+
+    max_iters : int
+        Max number of iterations for updating document topic distribution in E-step.
+
+    mean_change_tol : float
+        Stopping tolerance for updating document topic distribution in E-setp.
+
+    cal_diff : boolean
+        Parameter that indicate to calculate differene in `component_` or not.
+        Set `cal_diff` to `True` when we need to run M-step.
```

Comment:
I think the variable name `cal_diff` is misleading.
I will change it to `cal_sstats` since in the original code it is called [sstats](https://github.com/qpleple/online-lda-vb/blob/master/mdhoffma/onlineldavb.py#L189-L191).
And it corresponds to `sum( n_{dw} * phi_{dwk})` in M-step.